primitives/src/transaction.rs

```diff
@@ -349,6 +351,388 @@ impl Encodable for Transaction {
+                    if decoder.push_bytes(bytes)? {
+                        // Still more bytes required.
+                        return Ok(true);
+                    }
+                    let decoder = self.state.version_transition();
+                    let version = decoder.end()?;
+                    self.state = State::Inputs(version, Attempt::First, VecDecoder::<TxIn>::new());
+                }
+                State::Inputs(version, attempt, decoder) => {
+                    if decoder.push_bytes(bytes)? {
+                        return Ok(true);
+                    }
+                    // Copy the state because we need mutable access to self to transition.
+                    let version = *version;
+                    let attempt = *attempt;
+
+                    let decoder = self.state.inputs_transition();
+                    let inputs = decoder.end()?;
+
+                    if Attempt::First == attempt {
+                        if inputs.is_empty() {
+                            self.state = State::SegwitFlag(version);
+                        } else {
+                            self.state = State::Outputs(
+                                version,
+                                inputs,
+                                IsSegwit::No,
+                                VecDecoder::<TxOut>::new(),
+                            );
+                        }
+                    } else {
+                        self.state = State::Outputs(
+                            version,
+                            inputs,
+                            IsSegwit::Yes,
+                            VecDecoder::<TxOut>::new(),
+                        );
+                    }
+                }
+                State::SegwitFlag(version) => {
+                    if bytes.is_empty() {
+                        return Ok(true);
+                    }
+                    let version = *version;
+
+                    let segwit_flag = bytes[0];
+                    *bytes = &bytes[1..];
+
+                    if segwit_flag != 1 {
+                        return Err(E(Inner::UnsupportedSegwitFlag(segwit_flag)));
+                    }
+                    self.state = State::Inputs(version, Attempt::Second, VecDecoder::<TxIn>::new());
+                }
+                State::Outputs(version, _, is_segwit, decoder) => {
+                    if decoder.push_bytes(bytes)? {
+                        return Ok(true);
+                    }
+                    // These types are Copy, so we can deref them but does not work for vectors.
+                    let version = *version;
+                    let is_segwit = *is_segwit;
+
+                    // We get the inputs vector here instead of in the pattern match because I
+                    // couldn't find another way to get it out of behind the mutable reference.
+                    let (inputs, decoder) = self.state.outputs_transition();
```

Comment:
Would it be weird to have all the state transitions return variables like this and just shadow the match statement ones if required? Then we would avoid all the copy lines I think.

Reply:
Yep that would work, what is the value (other than its currently odd that we return `Clone` vars but not `Copy` vars)?

Can we create an issue for that to consider as a follow up please? The `TransactionDecoderState` is a public enum so we 100% have to make a call on this before 1.0

Reply:
BTW I'm pushing pretty hard on this encode/decode stuff to merge and follow up. You are 100% ok to push back, we are all big boys, if you feel strongly about it push hard bro.

Reply:
The transaction state is a beast, but I haven't thought of any alternative approaches I feel good about. Was just kinda tossin this idea out there as maybe a small clarity improvement, but not sure I am even convinced on it yet.

Reply:
No sweat.